### ISSUE TITLE
C10_WARP_SIZE breaks softmax/multinomial on RDNA 3.5 GPUs (gfx1151, warp_size=32)

### ISSUE THREAD

**Issue by cuttle-agent** [pytorch/pytorch#186229]
# C10_WARP_SIZE breaks softmax/multinomial on RDNA 3.5 GPUs (gfx1151, warp_size=32)

### Summary

On AMD RDNA 3.5 GPUs (gfx1151, e.g., Radeon 8060S / "Ryzen AI Max+" 395 integrated), `torch.softmax()` returns incorrect results because `C10_WARP_SIZE` resolves to 64 at runtime via the `warpSize` macro, while the hardware warp size is actually 32. This causes silent numerical corruption in any operation that relies on warp-level reductions — most critically softmax and multinomial sampling.

### Reproduction

```python
import torch
torch.cuda.init()
x = torch.randn(1, 1000, device='cuda')
s = torch.softmax(x, dim=-1).sum().item()
print(f"softmax sum: {s}")  # Expected: 1.0, Actual: ~0.5
assert s == 1.0, f"softmax sum={s}, expected 1.0 — C10_WARP_SIZE is wrong for this arch"
```

On an RDNA 3.5 GPU (gfx1151), this prints `softmax sum: 0.5` (or similar values << 1.0) instead of the expected `1.0`.

### Root Cause

In `c10/macros/Macros.h`, the ROCm path defines:

```c
#if defined(USE_ROCM)
#define C10_WARP_SIZE warpSize  // = 64 or 32 (Defined in hip_runtime.h)
#else
#define C10_WARP_SIZE 32
#endif
```

The comment says *"= 64 or 32"*, suggesting it's runtime-adaptive. However, `warpSize` in HIP is a **compile-time constant** that resolves to **64** for all ROCm targets — even RDNA GPUs where the actual hardware warp size is 32.

This means `C10_WARP_SIZE` is 64 on RDNA, but warp-level primitives like `WARP_SHFL_XOR(offset=WARP_SIZE/2=32)` cross the real 32-lane warp boundary, corrupting parallel reductions.

### Impact

- **Every RDNA 3.x GPU user** building PyTorch from source with `PYTORCH_ROCM_ARCH=gfx1151` (or any RDNA consumer GPU)
- Silent corruption: softmax returns wrong values, multinomial sampling is broken, training diverges or crashes with cryptic errors
- Affects inference too — Chatterbox TTS and other models that use `torch.multinomial()` for sampling produce garbage output
- This is the **#1 blocking bug** for RDNA consumer GPU users — it makes basic tensor operations subtly wrong

### Environment

- **GPU:** AMD Radeon 8060S (gfx1151, RDNA 3.5)
- **ROCm:** 6.4.3
- **PyTorch:** Built from source with `PYTORCH_ROCM_ARCH=gfx1151`
- **OS:** Ubuntu 26.04, kernel 7.0.0
- **HIP runtime:** System `libamdhip64.so.7` (via symlink from ROCm 6.4.3)

### Proposed Fix

**Option A (minimal, immediate):** Make `C10_WARP_SIZE` a runtime variable for ROCm, queried from the device:

```c
#if defined(USE_ROCM)
// warpSize is hardcoded to 64 in hip_runtime.h even on RDNA (actual warp_size=32)
// Use 32 as the safe default for consumer GPUs, or query at runtime
#define C10_WARP_SIZE 32
#else
#define C10_WARP_SIZE 32
#endif
```

**Option B (proper, longer-term):** Make `C10_WARP_SIZE` truly arch-aware by reading from `hipDeviceAttributeWarpSize` at runtime, or by using the `__gfx1151__` preprocessor define to select the correct value at compile time:

```c
#if defined(USE_ROCM)
  #if defined(__gfx1151__) || defined(__gfx1152__) || defined(__gfx1153__)
    #define C10_WARP_SIZE 32
  #else
    #define C10_WARP_SIZE 64
  #endif
#else
#define C10_WARP_SIZE 32
#endif
```

**Note:** Setting `C10_WARP_SIZE=32` even on CDNA GPUs (where actual warp_size=64) should be safe — it just means warp-level primitives use the lower 32 lanes instead of all 64, which is correct for the subgroup operations PyTorch actually uses. But arch-aware selection is preferable.

### Additional Context

**Official ROCm PyTorch wheels** also fail on RDNA 3.5 — they contain zero gfx1151 code objects, so every HIP kernel fails with `hipErrorNoBinaryForGpu`. Users must build from source, and building from source hits this `C10_WARP_SIZE` bug immediately.

**Related issues:**
- MIOpen workspace allocation also fails on gfx1151 (every conv reports "workspace required: N, provided size: 0") — workaround is `torch.backends.cudnn.enabled = False`
- `torch.compile` with `reduce-overhead` mode works well on RDNA once these bugs are worked around

### Workaround

Patch `c10/macros/Macros.h` before building:

```c
#if defined(USE_ROCM)
#define C10_WARP_SIZE 32  // RDNA 3.5 has warp_size=32
#else
#define C10_WARP_SIZE 32
#endif
```

Then rebuild PyTorch from source with `PYTORCH_ROCM_ARCH=gfx1151`.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @ROCmPyTorch

**Comment by xinyazhang** [comment]
Which PyTorch version are you referring to?
The host side C10_WARP_SIZE has been re-defined as `at::cuda::warp_size()` long time ago.

**Comment by h34v3nzc0dex** [comment]
@xinyazhang thanks — you're right, and I can confirm it's resolved on current ROCm.

I originally hit this on a from-source build against **ROCm 6.4.3**, where `c10/macros/Macros.h` still had `#define C10_WARP_SIZE warpSize` and `warpSize` compiled to a constant 64 even on gfx1151 — so the **device-side** warp reductions (softmax / multinomial) crossed the real 32-lane boundary.

On the current nightly — **PyTorch 2.11.0+rocm7.13.0a20260506** (git `11712d9f`), HIP 7.13 — the header is now the host/device split you're describing:

```c
#ifdef __HIPCC__
static inline int __host__ C10_WARP_SIZE_INTERNAL() { return at::cuda::warp_size(); }
static inline constexpr int __device__ C10_WARP_SIZE_INTERNAL() {
#if defined(__GFX9__)
  return 64;
#else
  return 32;
#endif
}
#endif
#define C10_WARP_SIZE (C10_WARP_SIZE_INTERNAL())
```

The `__device__ … __GFX9__ ? 64 : 32` branch is the part that actually fixes the kernels for RDNA (the host-side `at::cuda::warp_size()` alone wouldn't), and it's effectively the arch-aware "Option B" from the report.

Verified on a Radeon 8060S (gfx1151), same build, exact repro from the issue:

```
softmax sum n=1000/2048/8192/32768 : 1.000000     (was ~0.5)
multinomial empirical p            : [0.104, 0.204, 0.297, 0.395]   (target [0.1,0.2,0.3,0.4])
max|softmax_cpu - softmax_gpu|     : 9.3e-10
device warp_size                   : 32
```

So this is fixed on current ROCm — good to close from my side. Thanks for the pointer to the host-side refactor; happy to leave the gfx1151 repro up as a regression reference.
